NDTensors/src/lib/Sectors/src/abstractcategory.jl

```diff
@@ -0,0 +1,41 @@
+abstract type AbstractCategory end
+
+abstract type AbstractGroup <: AbstractCategory end
+
+label(c::AbstractCategory) = error("method `label` not defined for type $(typeof(c))")
+
+function dimension(c::AbstractCategory)
+  return error("method `dimension` not defined for category or group $(typeof(c))")
+end
+
+function fusion_rule(::Type{C}, l1, l2) where {C<:AbstractCategory}
+  return error("fusion_rule not defined for type $(C)")
+end
```

Comment:
Is this version needed? I was just picturing that types would overload `fusion_rule(c1::AbstractCategory, c2::AbstractCategory)`.

Reply:
This one would be good to discuss a bit. The reason for having this (Type, value1, value2) sort of interface is so that for the simpler fusion rules, like U1, there can be a very 'clean' and minimal implementation like e.g. for U1 and SU2:
```
fusion_rule(::Type{U1}, n1, n2) = (n1 + n2,)

fusion_rule(::Type{SU2}, j1, j2) = abs(j1 - j2):(j1 + j2)
```
where the reason for the Tuple in the first example is that the return value is that it needs to be something iterable (like a Tuple, UnitRange, etc.).

Otherwise the returned value needs to be a Vector of the same category type. So the above examples would become:
```
fusion_rule(a::U1, b::U1) = [U1(label(a)+label(b))]

function fusion_rule(s1::SU2, s2::SU2) 
  j1, j2 = label(s1), label(s2)
  return [SU2(J) for J in abs(j1-j2):(j1+j2)]
end
```

To be clear I am fine with just having the second type! An advantage is it would make it clearer what is needed to overload `fusion_rule` because each type would do it the same way. Also it might be slightly faster in some cases, which could matter when these get called thousands of times.

Reply:
One other very important thing we'll need to think through more is whether *either* of the above is actually enough, when "outer degeneracies" can occur. I can currently see three ways of dealing with those:

* Just repeat irreps that occur more than once. Then the outer degeneracy is just how many times an irrep repeats.
* Return the outer degeneracy along with the new irreps (fusion results) in some way, like by adding a degeneracy count inside the irrep (probably not) or returning pairs of `irrep => int` (I doubt this is a good idea though)
* Maybe the best way is to embrace the idea of fusion as acting not on irreps, but on graded spaces where in the math literature the spaces are usually assumed to have a degeneracy of 1. Then the fusion could return another graded space where now some of the blocks have degeneracies that can be larger than 1, i.e. outer degeneracies.

Reply:
Gotcha, thanks for the explanation.

Another possibility would be something in between, like:
```julia
label_fusion_rule(a::U1, b::U1) = label(a) + label(b)

function label_fusion_rule(s1::SU2, s2::SU2)
  j1, j2 = label(s1), label(s2)
  return abs(j1-j2):(j1+j2)
end
```
But I see why you designed it the way you did, since then at that point you may as well extract the label for the user (and I guess also the categories should always be the same if they are being fused, if so maybe that should be checked?). I would vote for calling the inner function defined on the labels something like `label_fusion_rule` to make it clear that it is fusing the labels.

Also interesting point about outer degeneracy. Maybe returning pairs isn't the worst idea? Though I agree at that point it may be better just to go fully with graded spaces. Perhaps the code could be designed such that if you return something that isn't a pair or graded space it converts it to a graded space with degeneracy of 1 automatically but I don't have a good picture in mind for how well that would work in practice.

Reply:
I like the `label_fusion_rule` idea. Looking back at the regular `fusion_rule` implementations, I guess they aren't so bad i.e. still pretty readable.

About the outer degeneracies, I'd be fine to sort of "punt" on it a bit and do something like return pairs for now which might end up being the best design. But then if it's not, I wouldn't yet view this PR, even if merged, as somehow fully committing us to the final overall design. We can always see how the whole fusion arrays system looks and revisit parts of it if needed in a future PR.

Reply:
Makes sense to me. I would focus on keeping the design here both general but also simple, so as long as we hit on both of those (and getting the code style right) I'm fine leaving some more advanced features like outer degeneracy for future PRs. Some of this will require trying things out in the wild, having Olivier test them out, etc. and iterating based on that.

Reply:
Great